### PR TITLE
[export] simplify automatic dynamic shapes processing

### PR DESCRIPTION
**Pull request by pianpwk**
Removing `_transform_shapes_for_default_dynamic` and `assume_static_by_default=False` as added in https://github.com/pytorch/pytorch/pull/133620.

This reverts back to `assume_static_by_default=True` with the use of dynamo decorators (e.g. `maybe_mark_dynamic, mark_static`, instead) for handling Dim.AUTO & Dim.STATIC instead. This is easier to maintain, as it doesn't requiring reasoning about "inverting" the dynamic_shapes specs, and also opens up usage of other decorators (`mark_dynamic, mark_unbacked`).

On the user side this change has no effect, but internally this means dynamic behavior is determined only by the `dynamic_shapes` specs (ignoring user-side input decorators following https://github.com/pytorch/pytorch/pull/135536), but transferring this information for _DimHints via decorators, for Dynamo/non-strict to create symbolic_contexts accordingly, e.g. https://github.com/pytorch/pytorch/blob/7c6d543a5b3d65d8f49420e60cda150faaa5b8a0/torch/_dynamo/variables/builder.py#L2646-L2666

One caveat is we don't raise errors for dynamic decorators on the user side, since we don't know if they're from user markings, or from re-exporting with inputs we've previously marked. 

Differential Revision: D63358628

### CODE REVIEW

File: torch/_export/non_strict_utils.py [hunk 1]
```diff
@@ -94,8 +93,14 @@ def fakify(
...
     n_dims = len(t.shape)
+    dynamic_sizes = [
+        DimDynamic.DYNAMIC
+        if i in getattr(t, "_dynamo_weak_dynamic_indices", {})
+        else DimDynamic.STATIC
+        for i in range(n_dims)
+    ]
     symbolic_context = StatelessSymbolicContext(
-        dynamic_sizes=[DimDynamic.DYNAMIC] * n_dims,
+        dynamic_sizes=dynamic_sizes,
```
Comment (avikchaudhuri) [inline torch/_export/non_strict_utils.py:103]: OK, trying to understand...so you start off with mostly STATIC here.

File: torch/_export/non_strict_utils.py [hunk 2]
```diff
@@ -136,10 +141,7 @@ def make_fake_inputs(
 
     combined_args = _combine_args(nn_module, args, kwargs)
     _check_dynamic_shapes(combined_args, dynamic_shapes)
-    transformed_dynamic_shapes = _transform_shapes_for_default_dynamic(
-        combined_args, dynamic_shapes
-    )
-    constraints = _process_dynamic_shapes(combined_args, transformed_dynamic_shapes)
+    constraints = _process_dynamic_shapes(combined_args, dynamic_shapes)
```
Comment (avikchaudhuri) [inline torch/_export/non_strict_utils.py:144]: Where does AUTO come in?

File: torch/export/_trace.py [hunk 3]
```diff
@@ -559,7 +551,8 @@ def _export_to_torch_ir(
             ), _ignore_backend_decomps():
                 gm_torch_level, _ = torch._dynamo.export(
                     f,
-                    dynamic_shapes=transformed_dynamic_shapes,  # type: ignore[arg-type]
+                    dynamic_shapes=dynamic_shapes,  # type: ignore[arg-type]
+                    assume_static_by_default=True,
```
Comment (avikchaudhuri) [inline torch/export/_trace.py:555]: What does this buy?

File: torch/export/dynamic_shapes.py [hunk 4]
```diff
@@ -1011,20 +888,43 @@ def update_symbols(path, tensor, shape):
         def _create_static_dim(tensor, i, value):
             return _StaticDim(str(value), (int,), {"value": value})
 
+        # clean out decorators from user side, or previous export call
```
Comment (avikchaudhuri) [inline torch/export/dynamic_shapes.py:891]: Interesting, so we'll only allow specs to dictate which ones are marked? Should we warn if any of these were set, asking to use specs instead?

File: torch/export/dynamic_shapes.py [hunk 5]
```diff
@@ -1064,10 +964,10 @@ def _get_dim_name_mapping(
             continue
         if isinstance(dim, int):
             continue
-        assert isinstance(dim, _Dim)  # dim hints should have boiled away
-        name_to_dim[dim.__name__] = dim
-        if isinstance(dim, _DerivedDim):
-            name_to_dim[dim.root.__name__] = dim.root  # type: ignore[attr-defined]
+        if isinstance(dim, _Dim):
```
Comment (avikchaudhuri) [inline torch/export/dynamic_shapes.py:967]: `else` what? Maybe handle that explicitly with a pass and then assert / give a useful error if anything unexpected happens?

**Review by avikchaudhuri — APPROVED** [review]
Looks fine as long as tests pass, but I'm not entirely sure what the effect of the change is in this PR, could you write a summary please?

What I'd like is the some kind of invariant to hold around STATIC, constant dims, DYNAMIC, AUTO, and None here. Looks like we're not transforming specs any more, but using a marking scheme to get what we want. Who consumes these markings, is it just dynamo, or is it make_fx as well?

**Comment by pianpwk** [comment]
> Looks fine as long as tests pass, but I'm not entirely sure what the effect of the change is in this PR, could you write a summary please?
> 
> What I'd like is the some kind of invariant to hold around STATIC, constant dims, DYNAMIC, AUTO, and None here. Looks like we're not transforming specs any more, but using a marking scheme to get what we want. Who consumes these markings, is it just dynamo, or is it make_fx as well?

Ah sorry, just realized the diff description didn't update the PR description too

File: torch/_export/non_strict_utils.py (hunk 1 above)
Comment (pianpwk) [reply torch/_export/non_strict_utils.py:103]: Yeah, for now everything here can be static except those marked with Dim.AUTO. This doesn't interfere with the old Dim() specs.

File: torch/_export/non_strict_utils.py (hunk 2 above)
Comment (pianpwk) [reply torch/_export/non_strict_utils.py:144]: DimHints now get handled in `_process_dynamic_shapes`

File: torch/export/_trace.py (hunk 3 above)
Comment (pianpwk) [reply torch/export/_trace.py:555]: Nothing functionality wise, we just don't have to maintain this specs inverting which is already pretty confusing to me today

File: torch/export/dynamic_shapes.py (hunk 4 above)
Comment (pianpwk) [reply torch/export/dynamic_shapes.py:891]: Ah added a comment in the last PR description line

**Comment by desertfire** [comment]
> @pytorchbot merge -f "unrelated lint"

@pianpwk , as a retrospect, the `sam_fast` failure was really and shouldn't been force merged.

**Comment by pianpwk** [comment]
> > @pytorchbot merge -f "unrelated lint"
> 
> @pianpwk , as a retrospect, the `sam_fast` failure was really and shouldn't been force merged.

@desertfire Ah sorry. I should've realized the test no longer timing out was related

**Pull request closed without merging**